### DISCUSSION TITLE
❌ deprecated [RFC] A deployment builder must always compile the project before the deployment

### DISCUSSION THREAD

**Discussion by JohannesHoppe** [angular-schule/ngx-deploy-starter#1] [Ideas]
Based on [Minkos statement](https://github.com/angular-schule/website-articles/pull/3#discussion_r315802100) I suggest that we all

1. always compile in `production` by default
    - or with another configuration by the use of the Option `--configuration`
2. and after that deploy to the target

Example:
https://github.com/angular-schule/ngx-deploy-starter/blob/463290b3c659f4e6f51d84d9a2ec0a4218760f1b/src/deploy/actions.ts#L25-L29

**TODO:**
Discuss / find a unified solution with all other maintainers of the existing deployment builders.

----

See also:
https://github.com/Azure/ng-deploy-azure/issues/23

FYI
@jamesdaniels @mgechev @johnpapa @santoshyadav198613 @shmool @huv1k @fmalcher

**Comment by santoshyadavdev** [comment]
I am also using the same, so the config can be overwritten, but a great initiative. As of now in our case, we are rebuilding every time before deploy.

I liked the option 3 from @johnpapa we can give users an option if they want to rebuild it.

**Comment by fmalcher** [comment]
I agree 100%! We can't make sure that developers always remember to build the app before deploying it. Thus, the build must be part of the deployment step to avoid deploying broken stuff that hangs around in our `dist`.
I'd like to see an opt-out via command-line option (`--no-build`?) which can be used when I'm sure I havent changed anything and want to deploy again.

tl;dr;

- Yes, always build before
- but give the opportunity to opt-out
- opt-out should not come through a wizard, but only as an explicit cmd option

**Comment by huv1k** [comment]
Our deployment builder always builds the source before actual deployment too. I find it natural to do it like this for the developer.

**Comment by JohannesHoppe** [comment]
Ok, @fmalcher so you are voting for `--configuration` AND `--no-build` (for the people who really, really know what they are asking for)?

I can agree to that.

**Comment by fmalcher** [comment]
@JohannesHoppe  Yes!

**Comment by JohannesHoppe** [comment]
Ladies and gentlemen: Lets agree to this:

- `--configuration` (alias `-c`): A named build target, as specified in the "configurations" section of angular.json
- `--no-build`: Skip build process during deployment

### State of Progress

* @huv1k from `@zeit/ng-deploy` added `--configuration` (alias `-c`) and `--no-build` in v0.2.1 👍   
* @santoshyadav198613 from `@netlify-builder/deploy` already added `--configuration` but not `--no-build`.
* I will add `--configuration` (alias `-c`) and `--no-build` for the next version of `angular-cli-ghpages`.

### Call for action

:x: [`@angular/fire`](https://github.com/angular/angularfire2) (@jamesdaniels  are you in?, see https://github.com/angular/angularfire2/issues/2063)
:x: [`@azure/ng-deploy`](https://github.com/Azure/ng-deploy-azure) (@shmool are you in?, see https://github.com/Azure/ng-deploy-azure/issues/23)
:ballot_box_with_check: [`@netlify-builder/deploy`](https://github.com/ngx-builders/netlify-builder) (@santoshyadav198613 are you in?, see https://github.com/ngx-builders/netlify-builder/issues/33)
:white_check_mark: [`@zeit/ng-deploy`](https://github.com/zeit/ng-deploy-now/) (@huv1k implemented this in v0.2.1)
:white_check_mark: [`angular-cli-ghpages`](https://github.com/angular-schule/angular-cli-ghpages) (implemented in v0.6.0)

**Comment by george43g** [comment]
It took me to actually go through some of the committed code to understand exactly what was going on - so the --configuration flag is being added here to determine which configuration is used to *build* the project, not to specify a deploy target.

I think that's a mistake. The `ng deploy` command might have several of its own configurations, in which case the behavior of the --configuration flag should allow the user to choose production, development, staging, qa, etc... - each configuration will then tell the deploy builder two things:
1. Which build target to use, which defines the build configuration (production, local, etc)
2. Which environment to deploy the project to (qa, staging, production, dev, etc...)

There may be various such combinations. 

I'm unsure wether angular cli doesn't recognise the --configuration flag by design or not. But because of this, this has allowed --configuration to be used as a custom option in one of the commits here. Meaning that if the CLI now starts to recognise --configuration in the same way it does with serve and build targets, it will break some of the above code :(

**Comment by george43g** [comment]
https://github.com/angular/angular-cli/issues/17332

Just a heads up, this bugfix might break the `--configuration` flag used by several of the above deployment builders.

**Comment by JohannesHoppe** [comment]
@george43g Thank you for your feedback. I fully agree with your objection. For my projects I have already renamed the option `configuration` to `buildTarget `. I will try to bring all other projects back in sync as well.

See
* https://github.com/angular-schule/angular-cli-ghpages/pull/120
* https://github.com/angular-schule/ngx-deploy-starter/pull/25
